src/analysis/normed_space/box_subadditive.lean

```diff
@@ -0,0 +1,541 @@
+two subboxes `[l, u] ∩ {x | x i ≤ c}` and `[l, u] ∩ {x | c ≤ x i}` is greater than or
+equal/equal/less than or equal to its value on `[l, u]`.
+
+The main result of this file is theorem `box_subadditive_on.eq_zero_of_forall_is_o_prod`. It says
+that `f l u = 0` provided that `f` is `box_subadditive_on` the interval `[l, u]`, `l ≤ u`, and
+for any `p ∈ [l, u]` we have `f l' u' = o(volume [l', u'])` as both `l'` tends to `p` along
+`[l, p]`, `u'` tends to `p` along `[p, u]`, and the subbox `[l', u']` is homothetic to `[l, u]`.
+-/
+
+variables {ι α β M : Type*}
+
+open set (univ pi Icc Ioi) function finset (hiding univ pi) filter
+open_locale big_operators topological_space nnreal filter
+
+/-!
+### Definitions and basic properties
+
+In this section we define `box_subadditive_on`, `box_additive_on`, and `box_supadditive_on`, and
+prove some basic properties.
+-/
+
+/-- A function `f : (ι → α) → (ι → α) → M` is called `box_subadditive_on` a set `s : set (ι → α)`
+if for any rectangular box `[l, u] ⊆ s` and a hyperplane `x i = c`, `l i ≤ c ≤ u i`,
+we have `f l u ≤ f' (I ∩ {x | x i ≤ c}) + f' (I ∩ {x | c ≤ x i})`, where `I = [l, u]`, and
+`f' [a, b]` means `f a b`. -/
+def box_subadditive_on [decidable_eq ι] [preorder α] [ordered_add_comm_monoid M]
+  (f : (ι → α) → (ι → α) → M) (s : set (ι → α)) :=
+∀ ⦃l u⦄ (hsub : Icc l u ⊆ s) ⦃m : ι → α⦄ (hm : m ∈ Icc l u) i,
+  f l u ≤ f l (update u i (m i)) + f (update l i (m i)) u
+
+/-- A function `f : (ι → α) → (ι → α) → M` is called `box_additive_on` a set `s : set (ι → α)`
+if for any rectangular box `[l, u] ⊆ s` and a hyperplane `x i = c`, `l i ≤ c ≤ u i`,
+we have `f' (I ∩ {x | x i ≤ c}) + f' (I ∩ {x | c ≤ x i}) = f l u`, where `I = [l, u]`, and
+`f' [a, b]` means `f a b`. -/
+def box_additive_on [decidable_eq ι] [preorder α] [has_add M] (f : (ι → α) → (ι → α) → M)
+  (s : set (ι → α)) :=
+∀ ⦃l u⦄ (hsub : Icc l u ⊆ s) ⦃m : ι → α⦄ (hm : m ∈ Icc l u) i,
+  f l (update u i (m i)) + f (update l i (m i)) u = f l u
+
+/-- A function `f : (ι → α) → (ι → α) → M` is called `box_supadditive_on` a set `s : set (ι → α)`
+if for any rectangular box `[l, u] ⊆ s` and a hyperplane `x i = c`, `l i ≤ c ≤ u i`,
+we have `f' (I ∩ {x | x i ≤ c}) + f' (I ∩ {x | c ≤ x i}) ≤ f l u`, where `I = [l, u]`, and
+`f' [a, b]` means `f a b`. -/
+def box_supadditive_on [decidable_eq ι] [preorder α] [ordered_add_comm_monoid M]
+  (f : (ι → α) → (ι → α) → M) (s : set (ι → α)) :=
+∀ ⦃l u⦄ (hsub : Icc l u ⊆ s) ⦃m : ι → α⦄ (hm : m ∈ Icc l u) i,
+  f l (update u i (m i)) + f (update l i (m i)) u ≤ f l u
+
+namespace box_subadditive_on
+
+variables [decidable_eq ι] [preorder α] [ordered_add_comm_monoid M] {s : set (ι → α)}
+  {f : (ι → α) → (ι → α) → M} {l m u : ι → α}
+
+protected lemma mono (h : box_subadditive_on f s) {t} (ht : t ⊆ s) : box_subadditive_on f t :=
+λ l u hsub, h (set.subset.trans hsub ht)
+
+lemma le_sum_finset_subboxes (h : box_subadditive_on f s) (hsub : Icc l u ⊆ s)
```

Comment:
docstring for this one? (It is not obvious at first sight that using `piecewise` cuts your rectangle in the corresponding subrectangles, whose number depends on the cardinality of `t`). (I see below that you have made this kind of docstring for the next lemma, you could essentially copy it here).